File: LeetCode/cpp/78.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int> &nums) {
        int n = nums.size();
        vector<vector<int>> ret;
        vector<int> subset;
        vector<int> visited(n, 0);
        BackTracking(nums, subset, ret, visited, 0, n);
        return ret;
    }

    void
    BackTracking(vector<int> &nums, vector<int> &subset, vector<vector<int>> &ret, vector<int> &visited, int i, int n) {
        ret.push_back(subset);
        if (i == n)
            return;
        for (int j = i; j < n; ++j) {
            if (visited[j] != 1) {
                visited[j] = 1;
                subset.push_back(nums[j]);
                BackTracking(nums, subset, ret, visited, j + 1, n);
                subset.pop_back();
                visited[j] = 0;
            }
        }
    }
};

File: LeetCode/python3/78.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        ret = []
        self.Backtrack(nums, [], ret, 0)
        return ret

    def Backtrack(self, nums, curr, ret, k):
        ret.append([c for c in curr])
        if k == len(nums):
            return curr
        for i in range(k, len(nums)):
            curr.append(nums[i])
            curr = self.Backtrack(nums, curr, ret, i + 1)
            curr = curr[:-1]
        return curr
